File: ebilet_react_typescript/src/models/Ticket.ts
import {type AppEvent, parseEvent } from "./Event";
import {type City, parseCity} from "./City.ts";
import {type District, parseDistrict} from "./District.ts";
import {parseVenue, type Venue} from "./Venue.ts";
import {parseUser, type User} from "./User.ts";

export interface TicketLocation {
    city_name?: string | null;
    cityName?: string | null;

    district_name?: string | null;
    districtName?: string | null;

    venue_name?: string | null;
    venueName?: string | null;

    venue_address?: string | null;
    venueAddress?: string | null;
}

export interface Ticket {
    id?: number | null;
    ticket_id?: number | null;
    ticketId?: number | null;

    event_id?: number | null;
    eventId?: number | null;

    event_title?: string | null;
    eventTitle?: string | null;

    ticket_code?: string | null;
    ticketCode?: string | null;

    qr_code_text?: string | null;
    qrCodeText?: string | null;

    price?: number | null;

    status?: string | null;

    ticket_status?: string | null;
    ticketStatus?: string | null;

    purchased_at?: string | null;
    purchasedAt?: string | null;

    used_at?: string | null;
    usedAt?: string | null;

    transaction_id?: string | null;
    transactionId?: string | null;

    event?: AppEvent | null;
    city?: City | null;
    district?: District | null;
    venue?: Venue | null;
    location?: TicketLocation | null;
    user?: User | null;

    result?: string | null;
}

export function parseTicketLocation(json: any): TicketLocation {
    return {
        city_name: json?.city_name ?? null,
        cityName: json?.city_name ?? null,

        district_name: json?.district_name ?? null,
        districtName: json?.district_name ?? null,

        venue_name: json?.venue_name ?? null,
        venueName: json?.venue_name ?? null,

        venue_address: json?.venue_address ?? null,
        venueAddress: json?.venue_address ?? null,
    };
}

export function parseTicket(json: any): Ticket {
    const id =
        json?.id === undefined || json?.id === null ? null : Number(json.id);

    const ticketId =
        json?.ticket_id === undefined || json?.ticket_id === null
            ? null
            : Number(json.ticket_id);

    const eventId =
        json?.event_id === undefined || json?.event_id === null
            ? null
            : Number(json.event_id);

    const price =
        json?.price === undefined || json?.price === null
            ? null
            : Number(json.price);

    return {
        id,

        ticket_id: ticketId,
        ticketId,

        event_id: eventId,
        eventId,

        event_title: json?.event_title ?? null,
        eventTitle: json?.event_title ?? null,

        ticket_code: json?.ticket_code ?? null,
        ticketCode: json?.ticket_code ?? null,

        qr_code_text: json?.qr_code_text ?? null,
        qrCodeText: json?.qr_code_text ?? null,

        price,

        status: json?.status ?? null,

        ticket_status: json?.ticket_status ?? null,
        ticketStatus: json?.ticket_status ?? null,

        purchased_at: json?.purchased_at ?? null,
        purchasedAt: json?.purchased_at ?? null,

        used_at: json?.used_at ?? null,
        usedAt: json?.used_at ?? null,

        transaction_id: json?.transaction_id ?? null,
        transactionId: json?.transaction_id ?? null,

        event: json?.event ? parseEvent(json.event) : null,
        city: json?.city ? parseCity(json.city) : null,
        district: json?.district ? parseDistrict(json.district) : null,
        venue: json?.venue ? parseVenue(json.venue) : null,
        location: json?.location ? parseTicketLocation(json.location) : null,
        user: json?.user ? parseUser(json.user) : null,

        result: json?.result ?? null,
    };
}

export function getResolvedTicketId(ticket: Ticket): number {
    return ticket.ticketId ?? ticket.ticket_id ?? ticket.id ?? 0;
}

File: EBiletSwift/EBiletSwift/models/Ticket.swift
//
//  Ticket.swift
//  EBiletSwift
//
//  Created by Alperen Saraç on 25.06.2026.
//

import Foundation

/*
    Ticket

    Bilet modelidir.

    Şu endpointlerde kullanılır:
    - ticket_buy.php
    - my_tickets.php
    - ticket_detail.php
    - ticket_check.php
*/
struct Ticket: Codable {

    /*
        Bazı endpointler id döndürebilir.
        Bazıları ticket_id döndürebilir.
        İkisini de destekliyoruz.
    */
    let id: Int?
    let ticketId: Int?

    let eventId: Int?
    let eventTitle: String?

    let ticketCode: String?
    let qrCodeText: String?

    let price: Double?

    /*
        active
        used
        cancelled
    */
    let status: String?
    let ticketStatus: String?

    let purchasedAt: String?
    let usedAt: String?
    let transactionId: String?

    let event: Event?
    let city: City?
    let district: District?
    let venue: Venue?
    let location: TicketLocation?
    let user: User?

    let checkedBy: CheckedBy?

    /*
        ticket_check.php sonucunda:
        approved
        already_used
        invalid
        cancelled
        passive_event
    */
    let result: String?

    enum CodingKeys: String, CodingKey {
        case id
        case ticketId = "ticket_id"
        case eventId = "event_id"
        case eventTitle = "event_title"
        case ticketCode = "ticket_code"
        case qrCodeText = "qr_code_text"
        case price
        case status
        case ticketStatus = "ticket_status"
        case purchasedAt = "purchased_at"
        case usedAt = "used_at"
        case transactionId = "transaction_id"
        case event
        case city
        case district
        case venue
        case location
        case user
        case checkedBy = "checked_by"
        case result
    }

    /*
        Kullanılabilir bilet id değerini tek yerden almak için yardımcı property.
    */
    var resolvedTicketId: Int? {
        return ticketId ?? id
    }
}

/*
    TicketLocation

    Backend bazı bilet listelerinde location objesi döndürüyor:

    {
        "city_name": "...",
        "district_name": "...",
        "venue_name": "...",
        "venue_address": "..."
    }
*/
struct TicketLocation: Codable {

    let cityName: String?
    let districtName: String?
    let venueName: String?
    let venueAddress: String?

    enum CodingKeys: String, CodingKey {
        case cityName = "city_name"
        case districtName = "district_name"
        case venueName = "venue_name"
        case venueAddress = "venue_address"
    }
}

/*
    CheckedBy

    QR kontrolünü yapan görevli bilgisi.
*/
struct CheckedBy: Codable {

    let id: Int
    let fullName: String

    enum CodingKeys: String, CodingKey {
        case id
        case fullName = "full_name"
    }
}
